image-set() should serialize in authored order

rreno opened the pull request:
#### 2fff06ab22a9a766d9314fdc4eaefeadbf67d63d
<pre>
image-set() should serialize in authored order
<a href="https://bugs.webkit.org/show_bug.cgi?id=251196">https://bugs.webkit.org/show_bug.cgi?id=251196</a>
rdar://104683422

Reviewed by Antti Koivisto.

To make choosing an image in the image-set more efficient we sort the set
by resolution. This was causing us to serialize in a different order than the value
was authored if the author did not also sort by resolution. This change preserves the order
of the image-set-options as authored while providing a vector of sorted indices to enable
linear lookups by device resolution.

In the case where the image-set is empty because there are no supported image types
we introduce the concept of an invalid image[0]. This is a Style object that represents
a zero-sized image. This way we can have a well-defined image object to use in the style,
rendering, and paint systems but no loads will be triggered and the invalid image will
have zero intrinsic size so as not to affect layout calculations.

[0] <a href="http://w3c.github.io/csswg-drafts/css-images-4/#invalid-image">http://w3c.github.io/csswg-drafts/css-images-4/#invalid-image</a>

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:
* LayoutTests/fast/css/image-set-type-crash-expected.txt: Added.
* LayoutTests/fast/css/image-set-type-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub.html:
    Added test cases for checking the computed style of an image-set with no supported image types.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSImageSetOptionValue.cpp:
(WebCore::CSSImageSetOptionValue::CSSImageSetOptionValue):
(WebCore::CSSImageSetOptionValue::create):
(WebCore::CSSImageSetOptionValue::equals const):
(WebCore::CSSImageSetOptionValue::customCSSText const):
(WebCore::CSSImageSetOptionValue::setResolution):
(WebCore::CSSImageSetOptionValue::setType):
(WebCore::CSSImageSetOptionValue::Type::Type): Deleted.
(WebCore::CSSImageSetOptionValue::Type::cssText const): Deleted.
(WebCore::CSSImageSetOptionValue::image const): Deleted.
(WebCore::CSSImageSetOptionValue::resolution const): Deleted.
(WebCore::CSSImageSetOptionValue::type const): Deleted.
* Source/WebCore/css/CSSImageSetOptionValue.h:
    Remove the CSSImageSetOptionValue::Type class. Since we no longer filter the image set
    by given type prior to making the StyleImageSet object, there&apos;s no need to check if the type is
    supported at this stage. Now we just hold the string as given to us by the CSS parser (if any).
    Since type is an optional parameter for the image set option, we represent the lack of
    a specified type with a null string.

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::createStyleImage const):

* Source/WebCore/rendering/style/StyleImage.h:
(WebCore::StyleImage::isGeneratedImage const):
(WebCore::StyleImage::isInvalidImage const):

* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::create):
(WebCore::StyleImageSet::StyleImageSet):
(WebCore::StyleImageSet::bestImageForScaleFactor):
* Source/WebCore/rendering/style/StyleImageSet.h:

* Source/WebCore/rendering/style/StyleInvalidImage.cpp: Copied from Source/WebCore/css/CSSImageSetOptionValue.h.
(WebCore::StyleInvalidImage::create):
(WebCore::StyleInvalidImage::StyleInvalidImage):
(WebCore::StyleInvalidImage::load):
(WebCore::StyleInvalidImage::image const):
(WebCore::StyleInvalidImage::computedStyleValue const):
* Source/WebCore/rendering/style/StyleInvalidImage.h: Copied from Source/WebCore/css/CSSImageSetOptionValue.h.

* Source/WebCore/rendering/style/StyleMultiImage.cpp:
(WebCore::StyleMultiImage::load):
* Source/WebCore/rendering/style/StyleMultiImage.h:

Canonical link: <a href="https://commits.webkit.org/261015@main">https://commits.webkit.org/261015@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe868c5a144c8a02fdd341fb0a8e9792bfdb8358

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110287 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19382 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42930 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1693 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119254 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114236 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20841 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10566 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102542 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116031 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/30384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85582 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12064 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/18031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7631 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14489 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/css/CSSImageSetValue.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "StyleBuilderState.h"
 #include "StyleImageSet.h"
 #include <wtf/text/StringBuilder.h>
+#include <numeric>
```
This is the Style failure. I think it's a false positive based on https://webkit.org/code-style-guidelines/#include-system

nt1m replied:
I'd just try to satisfy the style checker here, either by putting `<numeric>` before `<wtf...>` or by putting a new line before `<numeric>`

rreno replied:
It looks like in other places we consider WTF headers as system headers (hence the angle bracket include) and they are alphabetized along with STL headers.

I think I confused "system" for "standard library"

nt1m reviewed (commented):
Can you rebaseline GTK?

anttijk commented on Source/WebCore/rendering/style/StyleImageSet.h:
```diff
@@ -31,7 +31,7 @@ namespace WebCore {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    static Ref<StyleImageSet> create(Vector<ImageWithScale>);
+    static Ref<StyleImageSet> create(Vector<ImageWithScale>, Vector<size_t>);
```
why does this take Vectors while the constructor takes Vector&&?

rreno replied:
I was a bit torn here and I tried to follow the pattern I saw elsewhere. What I was going for was `create` would take either the vectors by value or you could pass a temporary like `StyleImageSet::create(WTFMove(v), WTFMove(u))`. Then the constructor will take by `Vector&&` so we can skip a copy by `WTFMove`-ing in the implementation of `create`.

Since the construction site of this object is only in once place in `CSSImageSetValue` and we're taking ownership of the vectors created there it's fine to make the create function take the arguments by rvalue reference I think.

anttijk commented on Source/WebCore/rendering/style/StyleImageSet.cpp:
```diff
@@ -81,8 +85,12 @@ ImageWithScale StyleImageSet::selectBestFitImage(const Document& document)
 {
-    ImageWithScale result;
-    for (auto& image : m_images) {
+    ImageWithScale result { StyleInvalidImage::create(), 1, String() };
```
Seems inefficient to construct StyleInvalidImage just to throw it away when an actual image is found. Can this be refactored to avoid it, (for example by using `std::optional<ImageWithScale>`)

rreno replied:
I could skip both the optional and the wasted construction of a `StyleInvalidImage` by checking if `ImageWithScale::image` is null or not after the loop. The `RefPtr` gives us a sort of implicit optional.

anttijk commented on Source/WebCore/css/CSSInvalidImageValue.h:
```diff
@@ -0,0 +1,55 @@
+    ~CSSInvalidImageValue() = default;
+
+    bool equals(const CSSInvalidImageValue&) const { return false; }
+    String customCSSText() const { ASSERT_NOT_REACHED(); return ""_s; }
```
why is this not reached?

rreno replied:
Oops accidentally left that in. I was checking whether or not we'd ever try to serialize this object.

anttijk commented on Source/WebCore/css/CSSInvalidImageValue.h:
```diff
@@ -0,0 +1,55 @@
+
+    ~CSSInvalidImageValue() = default;
+
+    bool equals(const CSSInvalidImageValue&) const { return false; }
```
why are two invalid images unequal?

rreno replied:
Since two invalid images are indistinguishable I had to choose whether or not to make all of them equal or none of them equal. I leaned slightly further toward this being a sort of "image NaN" and since `x != y` for all `x, y  := NaN` I went this route.

anttijk commented on Source/WebCore/css/CSSInvalidImageValue.h:
```diff
@@ -0,0 +1,55 @@
+namespace WebCore {
+
+// http://w3c.github.io/csswg-drafts/css-images-4/#invalid-image
+class CSSInvalidImageValue final : public CSSValue {
```
Why does this need a class (instead of, say adding an invalid state to CSSImageValue)?

rreno replied:
Adding an invalid state is probably better. I think a big sign this wasn't necessary was the fact that the whole thing could be represented trivially in a header only.

anttijk commented on Source/WebCore/rendering/style/StyleInvalidImage.h:
```diff
@@ -0,0 +1,64 @@
+
+class CSSInvalidImageValue;
+
+class StyleInvalidImage final : public StyleGeneratedImage {
```
Does this really need a whole class? (I'm suspicious of classes that do nothing).

rreno replied:
This class allows us to have a representation of an invalid image that is not a null pointer. This way we can avoid adding null checks (or `std::optional::has_value` or similar) elsewhere. Also since an invalid image is used in other places in the images level 4 spec I thought it would be useful to introduce this now so it can be used if and when we implement other parts of that spec.

I think the key thing this class gives us is an object that will return a null image of size zero so both rendering and paint can get well-defined values for intrinsic size and for image contents which match the spec's notion of an invalid image.

Maybe there's another way to do that too that I'm not aware of? I think this class is a straightforward way to achieve that though I'm open to alternatives.

rreno replied:
I agree that CSSInvalidImage does nothing though.

anttijk commented on Source/WebCore/css/CSSImageSetValue.cpp:
```diff
@@ -69,23 +70,19 @@ RefPtr<StyleImage> CSSImageSetValue::createStyleImage(Style::BuilderState& state
-        return a.scaleFactor < b.scaleFactor;
+    // We want to maintain the authored order for serialization so we create a sorted indexing vector.
+    Vector<size_t> sortedIndices(length);
+    std::iota(sortedIndices.begin(), sortedIndices.end(), 0);
```
that must be the weirdest function name is STL

rreno committed:
image-set() should serialize in authored order
https://bugs.webkit.org/show_bug.cgi?id=251196
rdar://104683422

Reviewed by Antti Koivisto.

To make choosing an image in the image-set more efficient we sort the set
by resolution. This was causing us to serialize in a different order than the value
was authored if the author did not also sort by resolution. This change preserves the order
of the image-set-options as authored while providing a vector of sorted indices to enable
linear lookups by device resolution.

In the case where the image-set is empty because there are no supported image types
we introduce the concept of an invalid image[0]. This is a Style object that represents
a zero-sized image. This way we can have a well-defined image object to use in the style,
rendering, and paint systems but no loads will be triggered and the invalid image will
have zero intrinsic size so as not to affect layout calculations.

[0] http://w3c.github.io/csswg-drafts/css-images-4/#invalid-image

* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:
* LayoutTests/fast/css/image-set-type-crash-expected.txt: Added.
* LayoutTests/fast/css/image-set-type-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-images/image-set/image-set-computed.sub.html:
    Added test cases for checking the computed style of an image-set with no supported image types.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSImageSetOptionValue.cpp:
(WebCore::CSSImageSetOptionValue::CSSImageSetOptionValue):
(WebCore::CSSImageSetOptionValue::create):
(WebCore::CSSImageSetOptionValue::equals const):
(WebCore::CSSImageSetOptionValue::customCSSText const):
(WebCore::CSSImageSetOptionValue::setResolution):
(WebCore::CSSImageSetOptionValue::setType):
(WebCore::CSSImageSetOptionValue::Type::Type): Deleted.
(WebCore::CSSImageSetOptionValue::Type::cssText const): Deleted.
(WebCore::CSSImageSetOptionValue::image const): Deleted.
(WebCore::CSSImageSetOptionValue::resolution const): Deleted.
(WebCore::CSSImageSetOptionValue::type const): Deleted.
* Source/WebCore/css/CSSImageSetOptionValue.h:
    Remove the CSSImageSetOptionValue::Type class. Since we no longer filter the image set
    by given type prior to making the StyleImageSet object, there's no need to check if the type is
    supported at this stage. Now we just hold the string as given to us by the CSS parser (if any).
    Since type is an optional parameter for the image set option, we represent the lack of
    a specified type with a null string.

* Source/WebCore/css/CSSImageSetValue.cpp:
(WebCore::CSSImageSetValue::createStyleImage const):

* Source/WebCore/rendering/style/StyleImage.h:
(WebCore::StyleImage::isGeneratedImage const):
(WebCore::StyleImage::isInvalidImage const):

* Source/WebCore/rendering/style/StyleImageSet.cpp:
(WebCore::StyleImageSet::create):
(WebCore::StyleImageSet::StyleImageSet):
(WebCore::StyleImageSet::bestImageForScaleFactor):
* Source/WebCore/rendering/style/StyleImageSet.h:

* Source/WebCore/rendering/style/StyleInvalidImage.cpp: Copied from Source/WebCore/css/CSSImageSetOptionValue.h.
(WebCore::StyleInvalidImage::create):
(WebCore::StyleInvalidImage::StyleInvalidImage):
(WebCore::StyleInvalidImage::load):
(WebCore::StyleInvalidImage::image const):
(WebCore::StyleInvalidImage::computedStyleValue const):
* Source/WebCore/rendering/style/StyleInvalidImage.h: Copied from Source/WebCore/css/CSSImageSetOptionValue.h.

* Source/WebCore/rendering/style/StyleMultiImage.cpp:
(WebCore::StyleMultiImage::load):
* Source/WebCore/rendering/style/StyleMultiImage.h:

Canonical link: https://commits.webkit.org/261015@main
